### ISSUE TITLE
API reference "latest" pages should yield HTTP status 301 not 404

### ISSUE THREAD

**Issue by chalin** [open-telemetry/opentelemetry-ruby#1435]
Visit one of the API reference pages and you'll initially get a 404, for example [opentelemetry-sdk/latest](https://open-telemetry.github.io/opentelemetry-ruby/opentelemetry-sdk/latest):

```console
$ curl -I https://open-telemetry.github.io/opentelemetry-ruby/opentelemetry-sdk/latest
HTTP/2 404 
```

If you [open the _same_ link in a browser](https://open-telemetry.github.io/opentelemetry-ruby/opentelemetry-sdk/latest), then **the page redirects** because it contains JS code that manipulates the window `location` to refer to the latest version of the appropriate Gem.

Since **API reference pages get _redirected_** the proper HTTP status returned should be 301, not 404. Otherwise, that's bad for SEO and for link checkers -- e.g., see https://github.com/open-telemetry/opentelemetry.io/pull/2505/files#r1135889989.

/cc @cartermp @svrnm

**Comment by arielvalentin** [comment]
I am not sure why it is hosted on `github.io` since it is available when we publish libraries here: https://www.rubydoc.info/gems/opentelemetry-sdk

**Comment by cartermp** [comment]
TIL! Let's update the docs site to point there for the API link too

**Comment by cartermp** [comment]
Submitted https://github.com/open-telemetry/opentelemetry.io/pull/2508

**Comment by cartermp** [comment]
I think we can consider this closed since we'll use the rubydoc link now.

**Comment by chalin** [comment]
We might have more to discuss, see:

- https://github.com/open-telemetry/opentelemetry.io/issues/2512

Also, having the same content in two places, isn't a good idea for SEO etc. My guess is that GH pages is used for dev versions of the API reference? If not, and it's the same as rubydoc.info, then maybe it's time to drop the GH pages?

At a minimum, the GH pages should all probably be marked [noindex, nofollow](https://developers.google.com/search/docs/crawling-indexing/block-indexing), and in addition have an appropriate `robots.txt` file.

**Comment by ahayworth** [comment]
The one nice thing that the GH pages version gives us is index page with links to all the docs: https://open-telemetry.github.io/opentelemetry-ruby/

It would be kinda nice to keep that somehow. But otherwise yeah - I don't really see a huge need to keep hosting these on GH pages. Does anyone else even do that?

That said - I'm really not sure exactly how these are getting generated today. I think it's a GitHub integration but because I'm only an approver on this repo, I can't actually see that. Maybe someone like @mwear with more admin privileges can check the settings pages and see if there's a github pages integration installed?

(For what it's worth, the custom code that redirects can be found [here on the gh-pages branch](https://github.com/open-telemetry/opentelemetry-ruby/blob/gh-pages/404.html))

**Comment by svrnm** [comment]
is this issue still a thing?

**Comment by chalin** [comment]
Yes it is. The URL in the opening comment still returns 404. I haven't investigated if the [same content in two places](https://github.com/open-telemetry/opentelemetry-ruby/issues/1435#issuecomment-1470411191) issue is still relevant.

**Comment by svrnm** [comment]
> Yes it is. The URL in the opening comment still returns 404. I haven't investigated if the [same content in two places](https://github.com/open-telemetry/opentelemetry-ruby/issues/1435#issuecomment-1470411191) issue is still relevant.

https://www.rubydoc.info/gems/opentelemetry-api
and
https://open-telemetry.github.io/opentelemetry-ruby/opentelemetry-api/v1.2.5/

still look very similar to me, so I guess there are still both places.

**Comment by svrnm** [comment]
> The one nice thing that the GH pages version gives us is index page with links to all the docs

We can have some of that on opentelemetry.io I guess:

* We can add links to the registry entries: https://opentelemetry.io/ecosystem/registry/?language=ruby
* We can add links to the appropriate documentation pages, e.g. exporters, instrumentation libraries, resource detectors, etc.

**Comment by chalin** [comment]
This is still relevant.

**Comment by chalin** [comment]
This is still relevant.

**Comment by svrnm** [comment]
apologies, I should not remove/add any labels, @open-telemetry/ruby-approvers can you unstale this/keep it until we have figured out how to resolve this?

**Comment by kaylareopelle** [comment]
I think we may be ready to pick this back up. GitHub pages have been in disarray for a while, and I think it's time we finally retired them.

**Comment by chalin** [comment]
> I think we may be ready to pick this back up. GitHub pages have been in disarray for a while, and I think it's time we finally retired them.

🙌🏻

**Comment by thompson-tomo** [comment]
@chalin / @svrnm as we now have the otel registry as our index (https://github.com/open-telemetry/opentelemetry.io/pull/10606) with links to all our docs, there is no need for the gh pages.

Are we able to turn off the github pages without any redirection to the new location given it wasn’t maintained/current? or should we do one of the below:
- establish if we can do a redirect to new url and after a period shut it down
- replace content with a message informing user of pending shutdown & after a month or so, shut it down.

Thanks for your input.

Cc @kaylareopelle

**Comment by chalin** [comment]
@thompson-tomo the registry solves discoverability, but this issue is narrower: the "latest" API-ref URLs on GH Pages still return a real 404 to non-browser clients (just re-checked, still 404), breaking inbound links and confusing link checkers/SEO.

@kaylareopelle's note above about retiring GH Pages would resolve this too: either drop the pages, or add a real server-side redirect while they're still up (what this issue asks for). I'd fold this into that retirement decision rather than close it on the registry argument alone. Of your three options, a temporary redirect-then-shutdown (option 1) directly satisfies this issue.

@kaylareopelle, where does the GH Pages retirement stand? Happy to close once it lands.

**Comment by thompson-tomo** [comment]
Thanks @chalin we had been awaiting for the registry entries to be published so that there was no longer any use case for a user to visit the gh pages. Hence we are now ready to retire them. I will look into adding the redirect and then shut it down after a while.

**Comment by thompson-tomo** [comment]
Pr #2287 available to perform the gh page retirement including redirects.

**Comment by chalin** [comment]
Closing as moot: opentelemetry.io no longer links into the GH Pages docs (verified: zero references to `open-telemetry.github.io/opentelemetry-ruby` across the site repo).

For any stray external links, #2287 replaces the dead 404s with redirect stubs pointing readers at the registry, which is a strict improvement. Note for the record: GitHub Pages can't emit real HTTP 301s, so non-browser clients will still see a 404 status; that's a platform limitation, and acceptable given mootness.

Thanks @thompson-tomo and @kaylareopelle!
